### PR TITLE
Query caching

### PR DESCRIPTION
**Pull request by fl0v**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #189

Basically i copied the relevant code from \yii\db\Connection and \yii\db\Command (same for the unit test).
I added the $db handler as a 4th param to `\yii\mongodb\Query::fetchRows()` to be able to generate the cache key.
This also fixes query profiling to take in consideration `$db->enableProfiling`.

I also changed ` Yii::info($token, __METHOD__);` to ` Yii::trace($token, __METHOD__);` to reduce the logging noise, if anyone thinks the cursor id is relevant information pls let me know to change it back.

I'm not sure my implementation of \yii\mongodb\Query::getCacheKey() is the most optimal or if it covers all cases.

### CODE REVIEW

**Comment by fl0v** [comment]
My mistake,
I just realized I implemented the cache layer after the cursor is executed, basically its useless.
I think i nead to modify \yii\mongodb\Command to consider the cache options before it builds the cursor...

**Comment by fl0v** [comment]
Unfortunetly the way Query and ActiveQuery work together makes it very hard to generate a corect cache key before the query is beeing executed.

In `Query::buildCursor()` the following happen:
- `ActiveQuery::prepare()` prepares `Query::$where`
- `ActiveQuery::getCollection()` prepares `Query::$from` and also builds Collection object,
-  The query is beeing executed and `\MongoDB\Driver\Cursor` is returned
There is no clear separation between the preparing stage and execution stage of a query :(

In conclusion for caching to work we nead to break BC and rethink the way ActiveQuery prepares Query properties before `Query::buildCursor()` is called.

Anyone has any other ideeas ?
I'll be away for about a week, when I get back I'll take a look at it again with fresh eyes, maybe i'll get a better picture then :)

**Comment by fl0v** [comment]
It seams an easy aproach was to extract `Query::prepare()` from `Query::buildCursor()` and call it before the cache check in each query type method (all, one,  scalar, column).

Not sure if it can still be considered backwards compatibile, tehnically normal usage is not changed but if anyone extended Query with different implementations for those methods or for buildCursor for that metter (its public), those implementations nead to be aware that `Query::prepare()` must be called explicitly.

**Comment by fl0v** [comment]
Also i'm not to happy with the practice of defaulting $db to Yii::$app->get('mongodb') if its null.
It was already in place so i just copied it but it should be passed as a property just as the other params (from, where, select, ...)
Maybe a different pull request for that...
